### PR TITLE
Sanitize argument-free object params before logging

### LINKED ISSUE #19772: Sanitize object params before they get logged from argument-free classes
### Description & Motivation

The motivation for this proposal is as follows: when you store classes (not-yet instantiated, but from the main file) in a module's hyperparameters to instantiate them later, the related entries in the dictionary are not sanitized.

### Pitch

For example, let's say my configuration is this:

```python
class Stepper():
    def __init__(self):
        self.scale = 3

    def step(self):
        return self.scale

config = {
    "stepper": Stepper
}
```

Then I want the hyperparameters that will be logged to look like this:

```python
config = {
    "criterion": "Stepper"
}
```

And not like this:

```python
config = {
    "criterion": "<__main__.Stepper object at 0x352255190>"
}
```

### Alternatives

When a module has at least one `__init__` argument, this problem doesn't exist:
```python
class Stepper:
    def __init__(self, scale):
        self.scale = scale

    def forward(self, x):
        return self.scale * x

config = {
    "stepper": Stepper,
    "config": {"scale": 0.0001}
}
```

Results in a logged configuration dictionary of form:

```python
config = {
    "criterion": "Stepper",
    "config": {"scale": 0.0001}
}
```

### Additional context

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by V0XNIHILI**
## What does this PR do?

This PR sanitizes object params before they get logged to WandB and Neptune.

Fixes #19772

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

I had fun coding :D

----
📚 Documentation preview 📚: https://pytorch-lightning--19771.org.readthedocs.build/en/19771/

### CODE REVIEW

**Comment by V0XNIHILI** [comment]
@awaelchli could you have a look at this PR? It's been open and ready for a while now and it is quite a simple change. Let me know what you think!

**Review by awaelchli — APPROVED** [review]
Looks great @V0XNIHILI 
Thanks for sending the fix.
I pushed minor changes (the test wasn't passing because you asserted the repr, but the repr would only apply to an instance anyway).
